File: mobile/src/logic/FFScraper.js
import axios from 'axios';

class FutbolFantasyScraper {
    constructor() {
        let proxy = "";
        if (typeof window !== 'undefined' && window.location && window.location.hostname === 'localhost') {
            proxy = "https://cors-anywhere.herokuapp.com/";
        }
        this.baseUrl = `${proxy}https://www.futbolfantasy.com`;
        this.teams = [
            "alaves", "athletic", "atletico", "barcelona", "betis", "celta", "elche",
            "espanyol", "getafe", "girona", "levante", "mallorca", "osasuna",
            "rayo-vallecano", "real-madrid", "real-oviedo", "real-sociedad",
            "sevilla", "valencia", "villarreal"
        ];
        this.headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
        };
    }

    async scrapeTeam(slug) {
        const url = `${this.baseUrl}/laliga/equipos/${slug}`;
        try {
            const res = await axios.get(url, { headers: this.headers, timeout: 15000 });
            const html = res.data;
            const playerProbs = {};

            // 1. Structural approach (split by card)
            const blocks = html.split('class="jugador');
            for (let i = 1; i < blocks.length; i++) {
                const chunk = blocks[i];
                const nameMatch = chunk.match(/class="nombre"[^>]*>([^<]+)<\/span>/) || 
                                 chunk.match(/alt="([^"]{3,40})"/) || 
                                 chunk.match(/>([^<]{3,40})<\/span>/);
                const probMatch = chunk.match(/probabilidad[^>]*>[\s\S]*?(\d+%)/);

                if (nameMatch && probMatch) {
                    const name = nameMatch[1].trim();
                    const probStr = probMatch[1].trim();
                    const pInt = parseInt(probStr.replace("%", ""));
                    if (name.length > 2 && !["Barcelona", "Madrid", "Atletico", "Getafe", "Villarreal", "Sevilla"].includes(name)) {
                        playerProbs[name] = probStr;
                    }
                }
            }

            // 2. Data-attribute fallback (if still missing some)
            const bits = html.split('data-probabilidad="');
            for (let i = 1; i < bits.length; i++) {
                const chunk = bits[i];
                const probStr = chunk.substring(0, chunk.indexOf('"'));
                let prob = probStr;
                if (prob === "Titular") prob = "100%";
                if (prob === "Suplente") prob = "0%";
                if (prob.includes("%") || prob === "100%") {
                    const searchArea = bits[i-1].slice(-1000);
                    const nameMatch = searchArea.match(/alt="([^"]{3,40})"/) || searchArea.match(/>([^<]{3,40})<\/span>/);
                    if (nameMatch) {
                        const name = nameMatch[1].trim();
                        const pInt = parseInt(prob.replace("%", ""));
                        const existing = playerProbs[name] ? parseInt(playerProbs[name].replace("%", "")) : -1;
                        if (pInt > existing && !["Barcelona", "Madrid", "Atletico", "Getafe", "Villarreal", "Sevilla"].includes(name)) {
                            playerProbs[name] = prob;
                        }
                    }
                }
            }
            return playerProbs;
        } catch (error) { return {}; }
    }

    async getAllLineupData() {
        console.log("[FFScraper] Starting ultra-resilient harvest...");
        const allProbs = {};
        for (const team of this.teams) {
            const teamProbs = await this.scrapeTeam(team);
            Object.assign(allProbs, teamProbs);
            await new Promise(r => setTimeout(r, 200));
        }
        return allProbs;
    }
}

export default FutbolFantasyScraper;

File: ff_scraper.py
import requests
from bs4 import BeautifulSoup
import re

class FutbolFantasyScraper:
    BASE_URL = "https://www.futbolfantasy.com"
    TEAMS = [
        "alaves", "athletic", "atletico", "barcelona", "betis", "celta", "elche",
        "espanyol", "getafe", "girona", "levante", "mallorca", "osasuna",
        "rayo-vallecano", "real-madrid", "real-oviedo", "real-sociedad",
        "sevilla", "valencia", "villarreal"
    ]

    def __init__(self):
        self.headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
        }

    def scrape_team_page(self, team_slug):
        url = f"{self.BASE_URL}/laliga/equipos/{team_slug}"
        try:
            response = requests.get(url, headers=self.headers, timeout=10)
            if response.status_code != 200: return {}
            soup = BeautifulSoup(response.content, "html.parser")
            player_probs = {}
            items = soup.find_all(attrs={"data-probabilidad": True})
            for item in items:
                prob = item["data-probabilidad"]
                if prob == "Titular": prob = "100%"
                elif prob == "Suplente": prob = "0%"
                name = None
                img = item.find("img", alt=True)
                if img: name = img["alt"]
                if not name:
                    name_tag = item.find(class_=re.compile(r"truncate-name|nombre", re.I))
                    if name_tag: name = name_tag.get_text(strip=True)
                if name: player_probs[name.strip()] = prob
            
            # Text fallback
            for tag in soup.find_all(string=re.compile(r"\d+%")):
                p = tag.strip()
                parent = tag.find_parent()
                if parent:
                    text = parent.get_text(" ", strip=True).replace(p, "")
                    m = re.search(r"([A-ZÁÉÍÓÚÑ][a-záéíóúñ]+(?:\s+[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+)+)", text)
                    if not m: m = re.search(r"([A-ZÁÉÍÓÚÑ][a-záéíóúñ]{3,})", text)
                    if m:
                        n = m.group(1).strip()
                        if n not in player_probs: player_probs[n] = p
            return player_probs
        except:
            return {}

    def get_all_lineup_data(self):
        print("Scraping FF data team by team...")
        # Dictionary of team_slug -> { player_name: prob }
        all_probs_by_team = {}
        for team in self.TEAMS:
            probs = self.scrape_team_page(team)
            all_probs_by_team[team] = probs
        return all_probs_by_team
